### DISCUSSION TITLE
question about Unlock-on-Polka

### DISCUSSION THREAD

**Discussion by fghdotio** [tendermint/tendermint#9246] [Q&A] [answered]
Hi, guys, I got a liveness question. I'll just quote the scenario described in ["Tendermint: Byzantine Fault Tolerance in the Age of Blockchains"](https://knowen-production.s3.amazonaws.com/uploads/attachment/file/1814/Buchman_Ethan_201606_Msater%2Bthesis.pdf):

_"Consider a round where A and B precommitted blockX while C and D pre-committed nil - a split vote. They all move to the next round, and blockY is proposed, which C and D prevote for. Suppose A is Byzantine, and prevotes for blockY as well (despite being locked on blockX), resulting in a polka. Suppose B does not see the polka and pre-commits nil, while A goes offline and C and D pre-commit blockY . They move to the next round, but B is still locked on blockX, while C and D are now locked on blockY , and since A is offline, they can never get a polka. Hence, we've compromised liveness with less than a third (here, only one) Byzantine validators."_

In such situation, B, locked on blockX, can only prevotes for blockX in the upcoming rounds, C and D can only prevote for blockY., meaning blockX, blockY, or any other new block can never get a polka. 

Let's say A, being Byzantine, stays offline forever. So how does the Unlock-on-Polka mechanism work since there's no way the now 3-node Tendermint cluster will get a polka in a higher round? As I see it, liveness is still compromised with only one Byzantine node.

Very confused here...

**Comment by cason** [comment, marked as answer]
In the reported scenario, the next proposer will either be locked on block X (B) or block Y (C, D). Therefore, it will re-proposed one of such blocks: X with valid round 0 or Y with valid round 1. Block X with valid round 0 cannot be accepted by C or D, as they locked Y in round 1. But block Y with valid round 1 us accepted by process B, provided it receives the prevote messages that made block Y to be locked by C and D at round 1.

**Reply by fghdotio** [reply]
_"... provided it receives the prevote messages that made block Y to be locked by C and D at round 1"_

Thanks! Feeling enlightened.

**Reply by fghdotio** [reply]
```go
// https://github.com/tendermint/tendermint/blob/v0.35.4/internal/consensus/state.go#L2112
if blockID, ok := prevotes.TwoThirdsMajority(); ok {
	// There was a polka!
	// If we're locked but this is a recent polka, unlock.
	// If it matches our ProposalBlock, update the ValidBlock

	// Unlock if `cs.LockedRound < vote.Round <= cs.Round`
	// NOTE: If vote.Round > cs.Round, we'll deal with it when we get to vote.Round
	if (cs.LockedBlock != nil) &&
		(cs.LockedRound < vote.Round) &&
		(vote.Round <= cs.Round) &&
		!cs.LockedBlock.HashesTo(blockID.Hash) {

		cs.Logger.Debug("unlocking because of POL", "locked_round", cs.LockedRound, "pol_round", vote.Round)

		cs.LockedRound = -1
		cs.LockedBlock = nil
		cs.LockedBlockParts = nil

		if err := cs.eventBus.PublishEventUnlock(cs.RoundStateEvent()); err != nil {
			return added, err
		}
	}
```
